lib/protocol2.js

```diff
@@ -0,0 +1,254 @@
+const Protomux = require('protomux')
+const safetyCatch = require('safety-catch')
+const b4a = require('b4a')
+const m = require('./messages')
+
+module.exports = class HypercoreProtocol {
+  constructor (mux, onunknowncore) {
+    this.muxer = Protomux.from(mux)
+    this.stream = this.muxer.stream
+
+    this.onunknowncore = onunknowncore || noop
+    this.closed = false
+
+    this._byDiscoveryKey = new Map()
+    this._byId = []
+    this._byRemoteId = []
+    this._opens = []
+
+    this.protocol = this.muxer.addProtocol({
+      context: this,
+      name: 'hypercore',
+      messages: 13,
+      onremoteclose: () => {
+        this.closed = true
+
+        const peers = this._byId
+
+        this._byId = []
+        this._byRemoteId = []
+        this._byDiscoveryKey.clear()
+
+        while (this._opens.length) {
+          this._opens.pop().done()
+        }
+
+        for (let i = 0; i < peers.length; i++) {
+          const peer = peers[i]
+          if (!peer) continue
+          this._callPeerClose(peer)
+        }
+      }
+    })
+
+    this.open = this.protocol.addMessage({ type: 0, encoding: m.wire.open, onmessage: onopen })
+    this.close = this.protocol.addMessage({ type: 1, encoding: m.wire.close, onmessage: onclose })
+    this.sync = this.protocol.addMessage({ type: 2, encoding: m.wire.sync, onmessage: onsync })
+    this.request = this.protocol.addMessage({ type: 3, encoding: m.wire.request, onmessage: onrequest })
+    this.cancel = this.protocol.addMessage({ type: 4, encoding: m.wire.cancel, onmessage: oncancel })
+    this.data = this.protocol.addMessage({ type: 5, encoding: m.wire.data, onmessage: ondata })
+    this.noData = this.protocol.addMessage({ type: 6, encoding: m.wire.noData, onmessage: onnodata })
+    this.want = this.protocol.addMessage({ type: 7, encoding: m.wire.want, onmessage: onwant })
+    this.unwant = this.protocol.addMessage({ type: 8, encoding: m.wire.unwant, onmessage: onunwant })
+    this.bitfield = this.protocol.addMessage({ type: 9, encoding: m.wire.bitfield, onmessage: onbitfield })
+    this.range = this.protocol.addMessage({ type: 10, encoding: m.wire.range, onmessage: onrange })
+    this.reorgHint = this.protocol.addMessage({ type: 11, encoding: m.wire.reorgHint, onmessage: onreorghint })
+  }
+
+  cork () {
+    this.protocol.cork()
+  }
+
+  uncork () {
+    this.protocol.uncork()
+  }
+
+  async _callPeerClose (peer) {
+    try {
+      await peer.onremoteclose()
+    } catch (err) {
+      safetyCatch(err)
+    }
+  }
+
+  _sync (id) {
+    return id < this._byRemoteId.length ? this._byRemoteId[id] : null
+  }
+
+  async _async (id) {
```

Comment:
`_async` and `_sync` are split as an optimization ya? Does it make sense to index the opens by ID (vs an array) to avoid dropping into `_async` if *any* peer is opening?

Reply:
Optimisation yes (super hot code). Unsure what you mean with the last part.

Reply:
code is gone now